extension/httpforwarder/extension.go

```diff
@@ -16,21 +16,101 @@ package httpforwarder
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
 
 	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
 )
 
 type httpForwarder struct {
+	forwardTo  *url.URL
+	httpClient *http.Client
+	server     *http.Server
+	logger     *zap.Logger
+	config     *Config
 }
 
 var _ component.ServiceExtension = (*httpForwarder)(nil)
 
-func (h httpForwarder) Start(_ context.Context, _ component.Host) error {
-	// TODO: Start HTTP server
+func (h *httpForwarder) Start(_ context.Context, host component.Host) error {
+	listener, err := h.config.Ingress.ToListener()
+	if err != nil {
+		return fmt.Errorf("failed to bind to address %s: %w", h.config.Egress.Endpoint, err)
+	}
+
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", h.forwardRequests)
+
+	h.server = h.config.Ingress.ToServer(handler)
+	go func() {
+		if err := h.server.Serve(listener); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
 	return nil
 }
 
-func (h httpForwarder) Shutdown(_ context.Context) error {
-	// TODO: Shutdown HTTP server
-	return nil
+func (h *httpForwarder) Shutdown(_ context.Context) error {
+	return h.server.Close()
+}
+
+func (h *httpForwarder) forwardRequests(writer http.ResponseWriter, request *http.Request) {
+	// Prepare new URL with value of final destination.
+	rawURL := fmt.Sprintf("%s://%s%s", h.forwardTo.Scheme, h.forwardTo.Host, request.RequestURI)
+	url, _ := url.Parse(rawURL)
+
+	forwarderRequest := request.Clone(context.Background())
+	forwarderRequest.URL = url
+	forwarderRequest.Host = url.Host
+	// Clear RequestURI to avoid getting "http: Request.RequestURI can't be set in client requests" error.
+	forwarderRequest.RequestURI = ""
+
+	// Add additional headers.
+	for k, v := range h.config.Egress.Headers {
+		forwarderRequest.Header.Add(k, v)
+	}
+
+	response, err := h.httpClient.Do(forwarderRequest)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusBadGateway)
+		return
+	}
+	defer response.Body.Close()
+
+	// Copy over response from the final destination.
+	writer.WriteHeader(response.StatusCode)
+	_, err = io.Copy(writer, response.Body)
+	if err != nil {
+		h.logger.Warn("Error writing HTTP response message", zap.Error(err))
+	}
```

Comment:
Why not just `response.Write(writer)`?

Reply:
I took a look at this. I noticed that the Status Code from response is not honored and also noticed that the response body is modified to include things like the Status Code from response and also the headers I believe. I think it would be better if the actual response body could be preserved. What do you think @bogdandrutu ?

An example where a server returns just `hello world` in its response body.

```
expected: "hello world"
actual  : "HTTP/1.1 202 Accepted\r\nContent-Length: 11\r\nContent-Type: text/plain; charset=utf-8\r\nDate: Mon, 14 Sep 2020 16:48:50 GMT\r\n\r\nhello world"
```

Reply:
Not an expert in the HTTP protocol, probably we should ask someone with more experience in this. @tigrannajaryan do you know about this?

Reply:
The docs suggest that the status will always be ok unless manually specified as done here: https://golang.org/pkg/net/http/#ResponseWriter.  Looks like this potentially derives a content type header if missing so the response may not necessarily be an exact copy, which could be problematic.

@asuresh4 are you sure we don't need to explicitly set all the headers as well?   Don't believe they come from the body.

Reply:
@rmfitzpatrick - yes, the current approach would require headers to copied over separately. Will address.

Reply:
Updated to copy over headers as well. Added `Via` header on both outgoing requests and responses based on @jrcamp's feedback.

@bogdandrutu - I chatted offline with @tigrannajaryan regarding the usage of `response.Write(writer)` he mentioned we shouldn't be sending headers within the body when relaying back to the client, which seems to be what it does. There could also be other undesired side effects like @rmfitzpatrick mentioned.